designs/intrinsics_design.md

```diff
@@ -0,0 +1,390 @@
+```
+
+This intrinsic function will resolve the condition in the list and then return the opposite boolean value returned. 
+
+### FN::Or
+
+```
+"Fn::Or": [{condition}, {...}]
+```
+
+This intrinsic function is very similar to the Fn::And function, but will check that at least one of the items in the last returns a truthy value. 
+
+## Pseudo Parameters
+
+Pseudo Parameters are predefined by AWS CloudFormation such as the AccountId and such. These will be resolved separately and is heavily used with Refs!. The result will cached in the properties dictionary so that there is conflict in there
+If the item is specified as an environment setting optionally or it is specified in the Paramater section it will be read there. Otherwise, These will be resolved whenever a !Ref Psuedo Paramater or Ref: Psuedo Paramater or ${Psuedo Paramater} is in the template.
+
+### AWS::AccountId
+
+This will be a static id used for local settings. 
+
+### AWs::NotificationArns
+
+This property can be randomly generated to follow the format
+
+### AWS::NoValue
+
+This will return the None python value.
+
+### AWS::Partition
+
+This is resource specific and seperates regions like US and China into subgroups. This will be found in a large dictionary.
+
+### AWS::REGION
+
+This property will first be attempted to be read from the environment settings. If it’s not there, it becomes randomly chosen from the list of regions.
+
+### AWS::StackId
+
+This will be a static id used for local settings. 
+
+### AWS::StackName
+
+This will be a static id used for local settings. 
+
+### AWS::URLSuffix
+
+This will be replaces by the corresponding [amazonaws.com](http://amazonaws.com/) or [amazonaws.com.cn](http://amazonaws.com.cn/) depending on the region settings. This will be found in a dictionary. 
+
+## Implementing the Code
+
+This can be separated into a IntrinsicsResolver and a SymbolResolver. The SymbolResolver is an abstraction that holds all the translation of the references and attributes inside the template. 
+The IntrinsicsResolver will recursively parse the template and all the attributes and the symbol table will be injected in at runtime. 
+
+### Intrinsic Symbol Table
+
+* logical_id_translator: This will be used as an exact translation for translating pseudo/logical_id types
+
+* default_type_resolver: This will be in the following form, resolving the type
+
+```
+{
+    "AWS::ApiGateway::RestApi": {
+        "RootResourceId": "/"
+    }
+}
+```
+
+* common_attribute_resolver: resolves common attributes that will be true across all
+
+```
+{    "Ref": lambda p,r: "",
+     "Arn:": arn_resolver}
+```
+
+
+First pseudo types are checked. If item is present in the logical_id_translator it is returned. Otherwise, it falls back to the default_pseudo_resolver. Then the default_type_resolver is checked, which has common attributes and functions for each types. Then the common_attribute_resolver is run, which has functions that are common for each attribute.
```

Comment:
All of this is for Pseudo Parameters? Pseudo parameters are all predefined. Seems like a lot of extra work that isn't needed.

Reply:
The paragraph might need to be clarified. The paragraph describes the flow of logic for resolving symbol table attributes.